services/galley/src/Galley/API/Internal.hs

```diff
@@ -438,29 +440,49 @@ sitemap = do
 
 rmUser :: UserId -> Maybe ConnId -> Galley ()
 rmUser user conn = do
-  let n = unsafeRange 100 :: Range 1 100 Int32
-  tids <- Data.teamIdsForPagination user Nothing (rcast n)
+  let n = toRange (Proxy @100) :: Range 1 100 Int32
+      nRange1000 = rcast n :: Range 1 1000 Int32
+  tids <- Data.teamIdsForPagination user Nothing n
   leaveTeams tids
-  cids <- Data.conversationIdRowsForPagination user Nothing (rcast n)
-  leaveConversations user cids
+  localDomain <- viewFederationDomain
+  allConvIds <- Query.conversationIdsPageFrom user (GetPaginatedConversationIds Nothing nRange1000)
+  goConvPages (toLocal (Qualified user localDomain)) nRange1000 allConvIds
   Data.eraseClients user
   where
+    goConvPages :: Local UserId -> Range 1 1000 Int32 -> ConvIdsPage -> Galley ()
+    goConvPages lusr range page = do
+      localDomain <- viewFederationDomain
+      let (remoteConvs, localConvs) = partitionRemoteOrLocalIds' localDomain . pageConvIds $ page
+      leaveLocalConversations localConvs
+      leaveRemoteConversations lusr remoteConvs
+      when (pageHasMore page) $ do
+        let usr = qUnqualified . unTagged $ lusr
+            nextQuery = GetPaginatedConversationIds (Just . pagePagingState $ page) range
+        newCids <- Query.conversationIdsPageFrom usr nextQuery
+        goConvPages lusr range newCids
     leaveTeams tids = for_ (Cql.result tids) $ \tid -> do
       mems <- Data.teamMembersForFanout tid
       uncheckedDeleteTeamMember user conn tid user mems
       leaveTeams =<< Cql.liftClient (Cql.nextPage tids)
-    leaveConversations :: UserId -> Cql.Page ConvId -> Galley ()
-    leaveConversations u ids = do
+    leaveLocalConversations :: [ConvId] -> Galley ()
+    leaveLocalConversations ids = do
       localDomain <- viewFederationDomain
-      cc <- Data.conversations (Cql.result ids)
+      cc <- Data.localConversations ids
       pp <- for cc $ \c -> case Data.convType c of
         SelfConv -> return Nothing
         One2OneConv -> Data.removeMember user (Data.convId c) >> return Nothing
         ConnectConv -> Data.removeMember user (Data.convId c) >> return Nothing
         RegularConv
+          -- TODO(md): see if this membership checking makes any sense here.
+          -- Perhaps we can simply call 'Update.removeMember' and let it handle
+          -- everything.
```

Comment:
Why would that be better? `removeMember` would query the whole conversation all over again, whereas here we already have a conversation value.

Reply:
To me this looks like duplicating the logic from `removeMember`, and it feels like checks are performed only partially. But I could be wrong.

Reply:
I'll drop this to-do item and leave the code as is; this is slightly out of scope of this PR.